[libpas] Implement pas_root_visit_conservative_candidate_pointers_in_address_range

Constellation opened the pull request:
#### 2d62c3414e1f589b6fe09124b2012c88012cfda8
<pre>
[libpas] Implement pas_root_visit_conservative_candidate_pointers_in_address_range
<a href="https://bugs.webkit.org/show_bug.cgi?id=248122">https://bugs.webkit.org/show_bug.cgi?id=248122</a>
rdar://102540311

Reviewed by Mark Lam.

This patch implements a function pas_root_visit_conservative_candidate_pointers_in_address_range. It takes a remote address and
size, then calling a callback for pointers in this region. This allows libpas to report customized pointer scheme including
compact pointers.

* Source/bmalloc/libpas/src/libpas/pas_platform.h:
* Source/bmalloc/libpas/src/libpas/pas_root.c:
(decode_system_compact_pointer):
(pas_root_visit_conservative_candidate_pointers_in_address_range):
* Source/bmalloc/libpas/src/libpas/pas_root.h:

Canonical link: <a href="https://commits.webkit.org/256874@main">https://commits.webkit.org/256874@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd886c13639dfe4dd60a1505c4deabface141c69

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97077 "17 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6344 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30195 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106594 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166865 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6579 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35074 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89466 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103282 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102746 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83693 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31946 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88002 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/367 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20123 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/83443 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/349 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21550 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28438 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4753 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5148 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/86151 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1590 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40865 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19416 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_root.h:
```diff
@@ -123,6 +123,18 @@ PAS_API extern pas_root* pas_root_for_libmalloc_enumeration;
+
+#define PAS_MAX_CANDIDATE_POINTERS 256
+
+typedef struct pas_conservative_candaite_pointer_and_location {
```
typo: /pas_conservative_candaite_pointer_and_location/pas_conservative_candidate_pointer_and_location/

Constellation replied:
Fixed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_root.h:
```diff
@@ -123,6 +123,18 @@ PAS_API extern pas_root* pas_root_for_libmalloc_enumeration;
+typedef struct pas_conservative_candaite_pointer_and_location {
+    vm_address_t address_of_pointer;
+    vm_address_t candidate_pointer;
+} pas_conservative_candaite_pointer_and_location;
```
Ditto.

Constellation replied:
Fixed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_root.h:
```diff
@@ -123,6 +123,18 @@ PAS_API extern pas_root* pas_root_for_libmalloc_enumeration;
+
+typedef kern_return_t (*pas_pointer_visitor_t)(void* context, const pas_conservative_candaite_pointer_and_location*, size_t num_candidates);
+
+PAS_API kern_return_t pas_root_visit_conservative_candidate_pointers_in_address_range(task_t task, void* context, vm_address_t zone_address, vm_address_t address, size_t size, memory_reader_t reader, pas_pointer_visitor_t visitor);
```
Should we call this `pas_root_visit_conservative_candidate_decoded_pointers_in_address_range` instead to make it clear that we won't do a normal pointer scan that does not require decoding?

Alternatively, should as also scan for and report candidate normal pointers?  The advantage of doing it ourselves is that we can apply more filtering (in the future) to rule out pointers that don't point back into libpas' heap.  The disadvantage is that we'll miss any pointers that may point out of libpas' heap, which I think exists?  What do you think?

Constellation replied:
I think we should keep the current name. So we may report decoded pointers, or may report the other pointers, but it depends on our implementation. And let leaks to do the current scanning + this function additionally. So we can have flexibility on what we will report from this function.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_root.c:
```diff
@@ -389,5 +390,101 @@ pas_root* pas_root_ensure_for_libmalloc_enumeration(void)
+            continue;
+
+        /* First page cannot be a pointer since it must be ummapped due to C standard and NULL */
+        if (candidate_pointer < (16ULL << 10))
```
IIRC, on Mac, the first 4G is unmapped.  Can you check if this is also true on other platforms, and filter out accordingly?

Constellation replied:
OK, filtered.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_root.c:
```diff
@@ -389,5 +390,101 @@ pas_root* pas_root_ensure_for_libmalloc_enumeration(void)
+            continue;
+
+        if (candidate_pointer < (4ULL << 10))
+            continue;
```
Ditto: double check that first 4G on x64 Mac is also unmapped, and apply appropriate filtering.

Constellation replied:
Done.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_root.c:
```diff
@@ -389,5 +390,101 @@ pas_root* pas_root_ensure_for_libmalloc_enumeration(void)
+    uintptr_t mapped_memory;
+    pas_root* root;
+    size_t number_of_candidates;
+    pas_conservative_candaite_pointer_and_location candidates[PAS_MAX_CANDIDATE_POINTERS];
```
Make sure to fix the typo in the typename here too.

Constellation replied:
Done.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_root.c:
```diff
@@ -389,5 +390,101 @@ pas_root* pas_root_ensure_for_libmalloc_enumeration(void)
+        address_of_pointer = address + (cursor - mapped_memory);
+        candidate_pointer = decode_system_compact_pointer(root, *((const uint32_t*)cursor));
+
+        if (!candidate_pointer)
+            continue;
```
This is redundant with the `candidate_pointer > FirstGuardPageSize` checks below, right?  I think we can remove it.

Constellation replied:
Yes, but I would like to keep it to make it explicit that we can skip pointer population if we return 0 from decode function. This branch is intentionally added to make it explicit.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_root.c:
```diff
@@ -389,5 +390,101 @@ pas_root* pas_root_ensure_for_libmalloc_enumeration(void)
+    /* Our address region is 36-bit, thus we encode a 16-byte aligned pointer into 32-bit value by shifting it by 4. */
+    return (vm_address_t)(((uintptr_t)value) << 4);
+#else
+    /* The other platforms are not supporting CompactPtr. */
+    PAS_UNUSED_PARAM(value);
```
It's a pity (and wasteful) to make `pas_root_visit_conservative_candidate_pointers_in_address_range` call `decode_system_compact_pointer` repeatedly when NOT `PAS_HAVE(36BIT_ADDRESS)`, only to never find any pointers.  Can you have `pas_root_visit_conservative_candidate_pointers_in_address_range` just return early if NOT `PAS_HAVE(36BIT_ADDRESS)` or define and check a `has_pointer_encoding_scheme_to_decode` boolean?  With the boolean, we can make it dependent on more than one encoding scheme in the future if we ever add other encoding schemes.

Constellation replied:
Done.

MenloDorian reviewed (approved):
r=me with fixes and improvements.

Constellation committed:
[libpas] Implement pas_root_visit_conservative_candidate_pointers_in_address_range
https://bugs.webkit.org/show_bug.cgi?id=248122
rdar://102540311

Reviewed by Mark Lam.

This patch implements a function pas_root_visit_conservative_candidate_pointers_in_address_range. It takes a remote address and
size, then calling a callback for pointers in this region. This allows libpas to report customized pointer scheme including
compact pointers.

* Source/bmalloc/libpas/src/libpas/pas_platform.h:
* Source/bmalloc/libpas/src/libpas/pas_root.c:
(decode_system_compact_pointer):
(pas_root_visit_conservative_candidate_pointers_in_address_range):
* Source/bmalloc/libpas/src/libpas/pas_root.h:

Canonical link: https://commits.webkit.org/256874@main
